core/search/revision_directory.py

```diff
@@ -0,0 +1,41 @@
+from typing import Mapping
+
+from core.config import CannotLoadConfiguration
+from core.search.revision import SearchSchemaRevision
+from core.search.v5 import SearchV5
+
+
+class SearchRevisionDirectory:
```

Comment:
Do we need anything more than the current and next expected index?

Reply:
Hopefully not, no. Replace `current` and `next` with their numeric values, and we get the same thing. :slightly_smiling_face:

Reply:
In case a downgrade is required, will deleting the previous highest schema work? (Ensuring the new highest is the rollback version)

Reply:
Downgrades definitely work as they're just an atomic alias change.

I've gone back and forth on adding a configuration option that allows for setting the schema version explicitly, but I'm still unsure about it. If an explicit schema version can be set in a configuration, that also means that at some point it needs to be unset. It's yet another manual bit of configuration for deployments. :thinking:

Reply:
Er, to be clear, I don't think deleting the highest schema is a good idea. It shouldn't break things, but it's a bit conceptually unsound; it essentially makes the system pretend that it doesn't know about a version.

Reply:
How would a downgrade work with it knowing all the versions, including the highest version?

Reply:
If the currently deployed version of the app knows about some set of versions `{ ... N, M}` and you're at version `M` and want to downgrade to `N`, it's safe to set `{base_name}-search-read` to `{base_name}-vN` as long as there exists a `{base_name}-vN-indexed` alias (that alias tells you that at some point index `N` was fully populated and working).

There's no explicit in-application way to do this right now as we actually never discussed one. I assumed it'd be a temporary emergency measure if some version turns out to be broken, so I assumed someone would be setting the value of the alias manually in opensearch.

If no `{base_name}-vN-indexed` alias exists, then that's going to involve repopulating the index. If we do need some kind of explicit in-app way to do this (explicitly set the version and repopulate the index) then I think that work is going to have to happen in a separate ticket. It's trivial to do, it's just that there's no exposed way to do it in the app right now.

Reply:
> It's trivial to do, it's just that there's no exposed way to do it in the app right now.

Trivial to do in that there's an API for this, and tests already that show how it works, but there's no exposed configuration setting to do it.

Reply:
I forgot to point out that deleting versions as a downgrade runs the risk of accidentally reusing a version number, and doing that makes the system unsound.

To illustrate:

1. I create some code for a new version `M`.
2. I run the CM and tell it to upgrade to version `M`.
3. I populate `M` with search documents.
4. I find out a few hours later that `M` is broken in some way, so I downgrade by setting the read/write pointers and delete the code for `M`.
5. I then write _new_ code and I reuse the version number `M`.
6. I deploy a new version of the CM and tell it to upgrade to `M`.
7. The CM sees that there's a populated index with version `M` and does nothing.

If instead, at step `5`, I used a new version number `P` where `P > M`, everything would have worked.

Version numbers cannot be reused!